torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py

```diff
@@ -65,6 +65,51 @@ def decompress(fut):
     return fut.then(decompress)
 
 
+class OptimizerHookState(object):
+    """
+    Holds state for running optimizer in-line after DDP communication hook.
+    Currently contains optimizer class as well as optimizer args.
+    """
+    __slots__ = ['functional_optim_cls', 'functional_optim_args', 'functional_optimizer']
+
+    def __init__(self, functional_optim_cls, *functional_optim_args):
+        # TODO: Support kwargs
+        self.functional_optim_cls = functional_optim_cls
+        self.functional_optim_args = functional_optim_args
+        self.functional_optimizer = self.functional_optim_cls(
+            [],
+            *self.functional_optim_args,
+            allow_empty_param_list=True
+        )
+        if not hasattr(self.functional_optimizer, 'step_param'):
+            raise ValueError(
+                f"Class {self.functional_optim_cls} must implement method step_param."
+            )
+
+def hook_then_optimizer(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future], optimizer_state: OptimizerHookState
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    """Runs optimizer in a functional fashion after DDP communication hook."""
+
+    def hook_then_optimizer_wrapper(
+        hook_state, bucket: dist.GradBucket
+    ) -> torch.futures.Future:
+        # Run original hook
+        fut = hook(hook_state, bucket)
```

Comment:
> will one separate CPU thread execute one hook and hold the future? is allreduce_work.wait() called inside hook() and fut value is filled inside hook() as well?

I think at this moment it is still just a future w/o completion for asynchronous. Once this hook/future appends a `.then(...)`, the successor future will wait for this work. Basically `wait()` is wrapped by `then`.

Reply:
`wait` will be called as normal in the reducer on the future that the hook returns, and yes, the future value is filled inside the user-installed communication hook as you mentioned. I guess the hook is run by one of the autograd threads, since we call `runHook` when a bucket is ready which happens in an autograd post hook.